modules/develop/pages/data-transforms/how-transforms-work.adoc

```diff
@@ -0,0 +1,41 @@
+= How Data Transforms Work
+:description: Learn how Redpanda data transforms work.
+
+Redpanda provides the framework to build and deploy inline transformations on data written to Redpanda topics, delivering processed and validated data to consumers in the format they expect. Redpanda does this directly inside the broker, eliminating the need to manage a separate stream processing environment or use 3rd-party tools. 
+
+image::shared:wasm1.png[Data transforms in a broker] 
+
+To learn how to build and deploy data transforms, see xref:./data-transforms.adoc[].
+
+== Wasm in Redpanda
+
+Data transforms use a WebAssembly (Wasm) engine inside a Redpanda broker, allowing Redpanda to control the entire transform lifecycle. For example, Redpanda can stop and start transforms when partitions are moved or to free up system resources for other tasks. 
+
+Within Repdanda, there's a single Raft controller that manages cluster information, including data transforms. On every shard (core), Redpanda knows what data transforms exist in the cluster, as well as metadata about the transform, such as input and output topics and environment variables. 
+
+Wasm is executed within a runtime in the broker. Data transforms take an input topic and map it to an output topic, running a virtual machine (VM) on the same shard as the input partition leader. Redpanda uses just-in-time (JIT) compilation, which compiles the function in memory, writes it to executable space, then runs the directly translated machine code. When it becomes leader of a given partition on the input topic, Redpanda does the following:
+
+. Spins up a Wasm VM using the JIT-compiled WebAssembly module
+. Pushes records from the input partition into the Wasm VM
+. Writes the output, which may go to the same shard on the same machine, or to a different shard on the same machine, or to a different machine
+
+image::shared:wasm_architecture.png[Wasm architecture in Redpanda]
+
+The transform function processes every record produced to an input topic and returns zero or more records that are then produced to an output topic. Data transforms are applied to all partitions on an input topic. A record is processed after it has been successfully written to disk on the input topic. Because the transform happens in the background after the write finishes, the transform doesn't affect the original produced record and doesn't block writes to the input topic. 
```

Comment:
A few questions that I think we should clarify here:

1. What happens when you deploy a new transform - does it process the input topic from the earliest or latest offset?
2. Does a transform track its offset, like any other consumer?
3. What happens when a partition leader moves to another broker? - how does it follow the leader and continue processing? This will also explain why transforms have at-least-once semantics.

Reply:
To answer:

> What happens when you deploy a new transform - does it process the input topic from the earliest or latest offset?

It starts from the latest offset. Deploys after resume from the previous offset.

> Does a transform track its offset, like any other consumer?

It's managed by the broker. It's not a "normal" consumer group, but we'll have lag metrics exposed.

> What happens when a partition leader moves to another broker? - how does it follow the leader and continue processing? This will also explain why transforms have at-least-once semantics.

When a partition loses leadership the broker internally issues a notification and we shutdown the transform on that core. The new broker also issues a notification when it gains leadership of a partition, which then the transform subsystem will start a processor and resume from the last committed offset.

> This will also explain why transforms have at-least-once semantics.
💯

Reply:
Incorporated. Please review the latest draft with fresh eyes to see how it all now reads. Should we move the section _Configure Wasm Properties_ to the **Run Data Transforms** page?

Reply:
> A new data transform processes the input topic from the latest offset. When a partition loses leadership, the broker issues an internal notification, and Redpanda shuts down the transform on that core. When another broker gains leadership of the partition, it also issues a notification, and the transform subsystem starts a processor and resumes from the last committed offset. This is why transforms have at-least-once semantics.

A new data transform will read the input topic from the latest offset (i.e. it will only read new data produced to the input topic). If a partition leader moves from one broker to another, then the data transform instance assigned to that partition must move with it. 

When a partition replica loses leadership, the broker hosting that partition replica stops the transform instance running on the same core. The broker that is now hosting the partitions new leader will start the transform on the same core as that leader, and the transform resumes from the last committed offset. 

If the previous transform instance failed to commit its latest offsets before moving with the partition leader (e.g. the broker crashed) then its likely that the new transform instance will reprocess some events. This is why transforms have at-least-once semantics.

Reply:
Thanks @jrkinley! Please confirm the latest draft!